shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -203,12 +203,18 @@ class ConstraintSolver extends NormalForms { self: Typer =>
                     case _ => reportError
                   }
               }
-            case (LhsRefined(N, ts, r), RhsBases(pts, N | S(L(_: FunctionType | _: TupleType)))) =>
-              reportError
-            case (LhsRefined(S(b: TupleType), ts, r), RhsBases(pts, S(L(ty: TupleType))))
-              if b.fields.size === ty.fields.size
-              => (b.fields.unzip._2 lazyZip ty.fields.unzip._2).foreach(rec(_, _, false))
-            case (LhsRefined(S(b: TupleType), ts, r), _) => reportError
+            case (LhsRefined(N, ts, r), RhsBases(pts, N | S(L(_: FunctionType | _: ArrayBase)))) => reportError
+            case (LhsRefined(S(b: TupleType), ts, r), RhsBases(pts, S(L(ty: TupleType)))) =>
+              if (b.fields.size === ty.fields.size)
+                (b.fields.unzip._2 lazyZip ty.fields.unzip._2).foreach(rec(_, _, false))
+              else
+                rec(b.toArray, ty.toArray, false)  // ?
+                // fail
+            case (LhsRefined(S(b: TupleType), ts, r), RhsBases(pts, S(L(ar: ArrayType)))) =>
+              rec(b.toArray, ar, false)
+            case (LhsRefined(S(ArrayType(i1)), ts, r), RhsBases(pts, S(L(ArrayType(i2))))) =>
+              rec(i1, i2, false)
```

Comment:
If you had a `toArray` method declared in `ArrayBase`, you could make this just one case:

```suggestion
            case (LhsRefined(S(b: ArrayBase), ts, r), RhsBases(pts, S(L(ar: ArrayType)))) =>
              rec(b.toArray, ar, false)
```